design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+
+The resolver will make use of `fieldpath` library we're using for composition to
+get and set the values unknown types and once it is done, an update request will
+be issued. In line with cross-resource references, if the field on the referenced
+object, pointed by `toFieldPath`, already has a value then the resolver will
+skip it, meaning it will resolve only once and users will need to delete that
+value for resolver to set it again.
+
+## Future Considerations
+
+### Transforms
+
+The retrieved value may not be in the form that's useful for the referencer
+resource directly and may need some formatting. If we see a great need for that,
+we can move the transform machinery from Crossplane to Crossplane Runtime and
+reuse it.
+
+### More Limited RBAC
+
+Currently, providers have access to all `Secret`s in the cluster. While this will
+provide a nice UX for generic reference users, it may not be desirable for security
+minded folks since it'd cause giving managed resource permission mean giving access
+to any `Secret` in the cluster. See [#2384](https://github.com/crossplane/crossplane/issues/2384).
+
+We can consider new settings in `Provider` or `ControllerConfig` to let users
+control this behavior with more granularity.
+
+### Composite Resources
+
+We could possibly have this on composite resources as well to keep them similar
+to managed resources, hence improving nested composition construction experience.
+But while its implementation would be straight-forward, there are some issues like
+whether we should let claim authors propagate this field as well and would that
+mean letting them access everything provider has access to, which is not few.
+
+### Obscure References
+
+There could be cases where you'd like to fetch a value from another resource but
+don't want to expose it on the managed resource object, i.e. make it available to
+controller to be inclduded in the calls but not visible on the CR. We could think
+of ways letting controller know that by either giving `toFieldPath: OBSCURED` or
+`makeAvailableAs: <some key in a map>`.
+
+## Alternatives Considered
+
+### Simpler API
+
+```yaml
+spec:
+  externalValues:
+    - group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fromFieldPath: spec.forProvider.cidrBlock
+      toFieldPath: spec.cidrBlock
+```
+
+We could certainly have gone with this approach as well but having everything at
+the same level increases the risk of having to break the API to add new features.
+Although it's probably safe that we will not see many new types of sources be
+added, because a lot of use cases are already covered by composition patch types,
+it could be desirable to have multiple source objects to construct a single value
+for example and that'd be a new type. So, we're taking the safer approach of
+letting the API be more open-ended for future developments and pay a small UX
+fee in exchange.
+
+### Multiple Source References
```

Comment:
I agree we don't want to build this out the gate, but history shows that it's definitely wise to structure our API such that we could cleanly add support in future if we needed to. :)

Reply:
Yeah I think we get that flexibility by putting the source under `fromObject` field instead of having it at the top. This way source fields can be added, only `toFieldPath` would be static which has been fine so far in composition.